Lib/asyncio/events.py

```diff
@@ -113,6 +113,34 @@ def _run(self):
             self._loop.call_exception_handler(context)
         self = None  # Needed to break cycles when an exception occurs.
 
+# _ThreadSafeHandle is used for callbacks scheduled with call_soon_threadsafe
+# and is thread safe unlike Handle which is not thread safe.
+class _ThreadSafeHandle(Handle):
+
+    __slots__ = ('_lock',)
+
+    def __init__(self, callback, args, loop, context=None):
+        super().__init__(callback, args, loop, context)
+        self._lock = threading.RLock()
+
+    def cancel(self):
+        with self._lock:
+            return super().cancel()
+
+    def cancelled(self):
+        with self._lock:
+            return super().cancelled()
+
+    def _run(self):
+        # The event loop checks for cancellation without holding the lock
+        # It is possible that the handle is cancelled after the check
+        # but before the callback is called so check it again after acquiring
+        # the lock and return without calling the callback if it is cancelled.
+        with self._lock:
+            if self._cancelled:
+                return
+            return super()._run()
```

Comment:
I don't think you want to run the callback with the lock held - you only need to load the callback and args

Reply:
it uses recursive mutex so is this an issue?

Reply:
An RLock doesn't work here because a signal handler can run that cancels the handle after the `if self._cancelled:` and before the `_run()`

Reply:
As I said on the issue, we are not guaranteeing signal safety here, signal handlers so call all APIs by call_soon_threadsafe, if they need to cancel handle then use `loop.call_soon_threadsafe(handle.cancel)`

Reply:
That's no use because the handle will have run by the time the call_soon_threadsafe runs the cancel

Reply:
i don't really understand why you want signal safety here so much, isn't this the same for all Python code? No Python code is safe if signal comes at unfortunate time,

Reply:
I'm gonna merge this now, I suggest you create a new issue regarding signal safety and provide code that reproduces the problem you are mentioning.